app/models/manager_refresh/inventory_collection/builder.rb

```diff
@@ -0,0 +1,234 @@
+      #   :auto_model_class
+      #     - tries to set model_class from persister class
+      #     - @see method auto_model_class
+      #   :without_model_class
+      #     - if false and no model_class derived or specified, throws exception
+      #   :adv_settings_enabled
+      #     - values from Advanced settings (doesn't overwrite values specified in code)
+      #     - see persister options() method
+      def self.default_options
+        {
+          :auto_object_attributes => true,
+          :auto_model_class       => true,
+          :without_model_class    => false,
+          :adv_settings_enabled   => true
+        }
+      end
+
+      # Entry point
+      # Creates builder and builds data for inventory collection
+      # @param name [Symbol] InventoryCollection.association value
+      #        (optional) method with this name also used for concrete inventory collection specific properties
+      # @param persister_class [Class] used for "guessing" model_class
+      # @param options [Hash]
+      def self.prepare_data(name, persister_class, options = {})
+        options = default_options.merge(options)
+        builder = new(name, persister_class, options)
+        builder.construct_data
+
+        yield(builder) if block_given?
+
+        builder
+      end
+
+      # @see prepare_data()
+      def initialize(name, persister_class, options = self.class.default_options)
+        @name = name
+        @persister_class = persister_class
+
+        @properties = {}
+        @inventory_object_attributes = []
+        @builder_params = {}
+        @dependency_attributes = {}
+
+        @options = options
+        @adv_settings = options[:adv_settings] # Configuration/Advanced settings in GUI
+        @shared_properties = options[:shared_properties] # From persister
+      end
+
+      # Builds data for InventoryCollection
+      # Calls method @name (if exists) with specific properties
+      # Yields for overwriting provider-specific properties
+      def construct_data
+        add_properties(:association => @name)
+        add_properties(:model_class => auto_model_class) if @options[:auto_model_class]
+
+        add_properties(@adv_settings, :missing) if @options[:adv_settings_enabled]
+        add_properties(@shared_properties, :missing)
+
+        send(@name.to_sym) if respond_to?(@name.to_sym)
+
+        add_inventory_attributes(auto_object_attributes) if @options[:auto_object_attributes]
+      end
+
+      # Creates InventoryCollection
+      def to_inventory_collection
+        if @properties[:model_class].nil? && !@options[:without_model_class]
+          raise MissingModelClassError
+        end
+
+        ::ManagerRefresh::InventoryCollection.new(to_hash)
+      end
+
+      #
+      # Missing method *add_some_property(value)* converted to:
+      # add_properties(:some_property => value)
+      #
+      def method_missing(method_name, *arguments, &block)
```

Comment:
do we use this somewhere? I think we could get rid of this...

I am fine with ```add_properties(:some_property => value)``` instead of ```add_some_property(value)```

Reply:
1) yes, we're using add_targeted_arel 

2) this is general interface for current and future needs. I think it makes sense to keep decision on developer which way (in some actually unknown situation) will look better (it means more readable)

Reply:
for others, I would rather have explicitely defined, which you set by ```add_properties``` and which by specific method